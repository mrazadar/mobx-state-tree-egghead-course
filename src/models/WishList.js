import { types, getParent, destroy } from "mobx-state-tree"

export const data = {
    "name": "Chronicales of Narnia Box Set - C.S. Lewis",
    "price": 28.73, 
    "image": "https://github.com/mobxjs/mobx-state-tree/blob/master/docs/mobx-state-tree-logo-gradient.png"
}

export const WishListItem = types
    .model ({
        name: types.string,
        price: types.number,
        image: types.optional(types.string, "") //or we can simply assign literal value empty string "", so mobx can figure out type for us. 
    })
    .actions(self => ({
        changeName(name){
            self.name = name
        },
        changePrice(price){
            self.price = price
        },
        changeImage(img){
            self.image = img
        },
        remove(){
            getParent(self, 2).remove(self)
        }
    }));

export const WishList = types
    .model({
        items: types.optional(types.array(WishListItem), [])
    })
    .actions(self => ({
        addItem(item) {
            self.items.push(item)
        },
        remove(item){
            // self.items.splice(self.items.indexOf(item), 1)
            destroy(item)
        }
    }))
    .views(self => ({
        get totalPrice(){
            return self.items.reduce((sum, entry)=> (sum + entry.price), 0)
        }
    }))