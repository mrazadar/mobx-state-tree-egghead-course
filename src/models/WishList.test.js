import { getSnapshot, onSnapshot, onPatch } from 'mobx-state-tree';
import { data, WishList, WishListItem } from './WishList';

import { reaction } from 'mobx';
import { list } from 'postcss';

it(`can create an instance of a WishListItem model.`, () => {
    const item = WishListItem.create(data);

    expect (item.price).toBe(28.73);
    item.changePrice(33)
    expect (item.price).toBe(33)
})


it(`can create an instance of a WishList model.`, () => {
    const wishList = WishList.create({
        items: [data]
    });

    expect (wishList.items.length).toBe(1);
    expect (wishList.items[0].price).toBe(28.73);
})



it(`can also add item in the WishList`, ()=> {
    
    const wishList = WishList.create();
    const states = [];
    const patches = [];

    onSnapshot(wishList, snapshot => {
        states.push(snapshot);
    });
    onPatch(wishList, patch => {
        patches.push(patch);
    });

    wishList.addItem(data);
    expect (wishList.items.length).toBe(1);

    wishList.addItem({
        name: 'Monogram', 
        price: 23,
    })
    expect (wishList.items.length).toBe(2);
    expect (wishList.items[1].price).toBe(23);
    wishList.items[1].changePrice(72.15);
    expect (wishList.items[1].price).toBe(72.15);

    expect (getSnapshot(wishList)).toEqual({
        items: [
            data, 
            {
                name: 'Monogram', 
                price: 72.15,
                image: ""
            }
        ]
    })
    expect (getSnapshot(wishList)).toMatchSnapshot()

    expect (states).toMatchSnapshot()

    expect (patches).toMatchSnapshot()
})


it(`can also add item in the WishList - patches`, ()=> {
    
    const wishList = WishList.create();
    const patches = [];

    onPatch(wishList, patch => {
        patches.push(patch);
    });

    wishList.addItem(data);
    expect (wishList.items.length).toBe(1);

    wishList.addItem({
        name: 'Monogram', 
        price: 23,
    })
    wishList.items[1].changePrice(72.15);    
    expect (patches).toMatchSnapshot()
})

it(`can also test views with computed properties`, ()=> {
    const wishList = WishList.create({
        items: [
            data,
            {
                name: 'Monogram', 
                price: 72.15,
                image: ""
            }
        ]
    });
    expect(wishList.totalPrice).toBe((72.15+data.price))
    let changed = 0;
    reaction(() => wishList.totalPrice, () =>  changed++)
    expect(changed).toBe(0)
    wishList.items[0].changeName("Test");
    expect(changed).toBe(0)
    wishList.items[0].changePrice(3)
    expect(changed).toBe(1)
})