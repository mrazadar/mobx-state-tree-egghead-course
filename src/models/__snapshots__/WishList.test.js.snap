// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can also add item in the WishList - patches 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "https://github.com/mobxjs/mobx-state-tree/blob/master/docs/mobx-state-tree-logo-gradient.png",
      "name": "Chronicales of Narnia Box Set - C.S. Lewis",
      "price": 28.73,
    },
  },
  Object {
    "op": "add",
    "path": "/items/1",
    "value": Object {
      "image": "",
      "name": "Monogram",
      "price": 23,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/1/price",
    "value": 72.15,
  },
]
`;

exports[`can also add item in the WishList 1`] = `
Object {
  "items": Array [
    Object {
      "image": "https://github.com/mobxjs/mobx-state-tree/blob/master/docs/mobx-state-tree-logo-gradient.png",
      "name": "Chronicales of Narnia Box Set - C.S. Lewis",
      "price": 28.73,
    },
    Object {
      "image": "",
      "name": "Monogram",
      "price": 72.15,
    },
  ],
}
`;

exports[`can also add item in the WishList 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "https://github.com/mobxjs/mobx-state-tree/blob/master/docs/mobx-state-tree-logo-gradient.png",
        "name": "Chronicales of Narnia Box Set - C.S. Lewis",
        "price": 28.73,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "https://github.com/mobxjs/mobx-state-tree/blob/master/docs/mobx-state-tree-logo-gradient.png",
        "name": "Chronicales of Narnia Box Set - C.S. Lewis",
        "price": 28.73,
      },
      Object {
        "image": "",
        "name": "Monogram",
        "price": 23,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "https://github.com/mobxjs/mobx-state-tree/blob/master/docs/mobx-state-tree-logo-gradient.png",
        "name": "Chronicales of Narnia Box Set - C.S. Lewis",
        "price": 28.73,
      },
      Object {
        "image": "",
        "name": "Monogram",
        "price": 72.15,
      },
    ],
  },
]
`;

exports[`can also add item in the WishList 3`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "https://github.com/mobxjs/mobx-state-tree/blob/master/docs/mobx-state-tree-logo-gradient.png",
      "name": "Chronicales of Narnia Box Set - C.S. Lewis",
      "price": 28.73,
    },
  },
  Object {
    "op": "add",
    "path": "/items/1",
    "value": Object {
      "image": "",
      "name": "Monogram",
      "price": 23,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/1/price",
    "value": 72.15,
  },
]
`;
